doc/developer/design/20210413_source_sink_resource_sharing.md

```diff
@@ -0,0 +1,269 @@
+database. If you were to instantiate a PostgreSQL source twice in Materialize,
+you'd wind up streaming two separate copies of the entire database to
+Materialize, which is certain to be surprising to users.
+
+With dynamic resource sharing, creating a new instance of a PostgreSQL source
+would instead backfill the missing data for just the one table, then cut over to
+the already existing replication stream. This will be cheap enough that we'll
+be comfortable allowing PostgreSQL sources to be unmaterialized.
+
+### Discussion
+
+The major downside of this approach is the implementation complexity. Cutting
+source instances over from a backfill resource to the steady-state resource will
+require a lot of fiddly logic and testing.
+
+My contention is that this complexity is well worth it for the user experience
+it provides. Expecting users to predict exactly which fields they'll need from a
+source is unrealistic. Asking them to recreate an entire stack of views, as
+described in the [workaround](#todays-workaround) is acceptable in the short
+term but unfriendly in the long term. But letting users interactively create the
+sources and views they want, while converging on the optimal implementation
+nonetheless? That's the magic experience we strive for.
+
+Also, the lessons learned here will be directly applicable to our future plans
+for multi-query optimization. Dynamically sharing resources in sources is a
+simplified version of dynamically sharing resources (i.e., arrangements) between
+views.
+
+## Alternatives
+
+### Bulk creation
+
+We'd previously talked about using SQL transactions to permit bulk creation
+of set of sources and views, like so:
+
+```sql
+BEGIN;
+CREATE SOURCE kafka_src FROM KAFKA BROKER '...' TOPIC 'top' FORMAT AVRO ...
+CREATE MATERIALIZED VIEW view1 AS SELECT col1 FROM kafka_src
+CREATE MATERIALIZED VIEW view2 AS SELECT col2 FROM kafka_src
+COMMIT;
+```
+
+When the transaction commits, Materialize could look holistically at the
+objects within, and realize that only one new source instance is necessary.
+
+But this turns out not to be any more flexible than the [current
+workaround](#todays-workaround). You need to be willing to declare all of your
+sources and views in one shot, and that's not always feasible.
+
+We might, however, want to support this feature for savvy users though as a
+minor optimization for dynamic resource sharing. Dynamic resource sharing only
+guarantees that the system converges to an optimal state. Creating sources one
+by one will result in a period of inefficiency, as several backfill resources
+are created and then merged together. But if the user is willing to provide a
+batch of create statements simultaneously, we can create the optimal number of
+instances from the start.
+
+### Schema changes
+
+In theory, if we supported arbitrary schema changes on live views, we wouldn't
+need resource sharing. In practice we are at least several years away from
```

Comment:
Does my example earlier of a set of views that can't be efficiently supported by an upstream materialized view make this claim not actually true?

(In reality I think it might be the other way around: schema changes will require some backfill mechanism, so building out dynamic sharing might help us start to navigate that design space a bit)